[WIP] feat(wandb): add logging to TFTrainer

borisdayma opened the pull request:
Bring logging feature parity from `Trainer` to `TFTrainer`.
Code has been refactored to share logging utilities.

borisdayma commented:
Note that there are still a few differences.

For example `TFTrainer` uses `args.eval_steps`.
It could make sense to share refactor training args so they share the same ones when possible.

borisdayma commented:
Here is an example run with `TFTrainer` on MRPC.

![image](https://user-images.githubusercontent.com/715491/83705691-3db64e00-a5db-11ea-93d1-c51a83559997.png)

[Link to W&B run](https://app.wandb.ai/borisd13/huggingface/runs/1zngxzw0?workspace=user-borisd13)

jplu commented on src/transformers/trainer_utils.py:
```diff
@@ -1,6 +1,83 @@
+    if iterator is not None:
+        iterator.write(output)
+    else:
+        print(output)
```
The logger should be used here instead of a print.

borisdayma replied:
Noted, this section comes from `Trainer` but I agree that it makes more sense to use the logger.

jplu commented on src/transformers/trainer_utils.py:
```diff
@@ -1,6 +1,83 @@
+    if trainer.epoch is not None:
+        logs["epoch"] = trainer.epoch
+    if trainer.tb_writer:
+        if trainer.__class__.__name__ == "Trainer":
+            for k, v in logs.items():
+                trainer.tb_writer.add_scalar(k, v, trainer.global_step)
+        elif trainer.__class__.__name__ == "TFTrainer":
+            import tensorflow as tf
+
+            with trainer.tb_writer.as_default():
+                for k, v in logs.items():
+                    tf.summary.scalar(k, v, step=trainer.global_step)
```
IMO these parts should belong to their respective Trainer.

borisdayma replied:
I see your point.
This is only related to the tensorboard logging so I could have a `trainer.tb_log` function associated to each `Trainer` and `TFTrainer` class.

jplu replied:
Very good idea!

jplu commented on src/transformers/trainer_utils.py:
```diff
@@ -1,6 +1,83 @@
+    trainer._wandb = wandb
+
+
+def log_metrics(trainer, logs: Dict[str, float], iterator: Optional[tqdm] = None) -> None:
```
Epoch and step might be passed as arguments which doesn't make too changes in the Trainer classes.

borisdayma replied:
Here it was mainly to keep it the same way as `Trainer`. There are a few advantages in doing it this way:

* some example scripts call for evaluation after full training (ex: `run_glue`), where the epoch & step are not accessible any longer, which were the reasons why we added them within the class.
* access to `trainer.tb_log` discussed in your previous comment will also be easier this way

jplu replied:
For evaluation I don't think you need the epoch. For the step yes indeed. But try to do with the `trainer.tb_log` it should be easier for me to see what you mean :)

jplu commented on src/transformers/trainer_tf.py:
```diff
@@ -273,45 +284,44 @@ def train(self) -> None:
         logger.info("  Total optimization steps = %d", self.train_steps)
 
+        self.global_step = 0
+        self.epoch = 0
```
Can you elaborate more on why this variable plz? Why not taking the epoch value just above.

borisdayma replied:
This is a float value showing number of epochs elapsed.
When comparing training runs of different batch size, or different data size, you will have different number of steps. You can plot loss vs epoch and superpose multiple runs this way.

![image](https://user-images.githubusercontent.com/715491/83760909-3888eb80-a63b-11ea-9e95-57bc1324d1cf.png)

jplu replied:
OH I see!!! I didn't know that (sorry I'm not really familiar with wandb) it ok makes sense.

jplu replied:
Would be nice if you can give a more elaborate name instead of just `epoch` (to do not confuse with the loop epoch variable) and add a comment explaining that it is used only for wandb.

borisdayma replied:
Ok, I just kept the same name as the one used in `Trainer`.
Any suggestion? `epoch_elapsed` would also bring confusion.
Another possibility would be to change the loop epoch to `epoch_iter`. What do you think?

jplu replied:
I suggest `epoch_logging` and indeed `epoch_iter` changing both + a small comment for `epoch_logging`

jplu commented on src/transformers/trainer_tf.py:
```diff
@@ -273,45 +284,44 @@ def train(self) -> None:
         logger.info("  Num Epochs = %d", epochs)
         logger.info("  Total optimization steps = %d", self.train_steps)
 
+        self.global_step = 0
```
This variable seems a repetition with the `iterations` variable can be removed.

borisdayma replied:
This is to make it compatible with the functions moved from `Trainer` to `TFTrainer`.
Could `iterations` be removed or does it capture something more complex (such as reloading from checkpoints…).

jplu replied:
It captures the step directly from the checkpoint if any so, it is a real duplicate of global.

borisdayma commented:
@jplu while I'm doing it, is the `step` variable from `_prediction_loop` used or can I remove it?

jplu commented:
It is used for logging.

jplu commented:
Ping @julien-c as there are some changes in the PT trainer.

borisdayma commented:
> It is used for logging.

Ok, I don't see it being used anywhere in that function…

jplu commented:
Hummm indeed... Can you still keep it, I will review this later to better check that part :)

borisdayma commented:
I should have addressed your comments. I moved tensorboard specific logging directly within specific trainers and call it from shared `log_metrics`.
It could have also been the opposite with a public `trainer.log` method that calls `log_metrics` for everything non-tensorboard (wandb and stdout) but I thought this separation would be less obvious to users.

Let me know if you have any other suggestions.

The main differences now between the 2 trainers are the use of `args.debug` and `args.eval_steps` in `TFTrainer`.

borisdayma commented:
Let me know if anything else is needed

jplu commented on src/transformers/trainer.py:
```diff
@@ -287,7 +287,7 @@ def get_optimizers(
         or override this method in a subclass.
         """
-        if self.optimizers is not None:
+        if self.optimizers is not Noneself.eval:
```
Be careful here!

borisdayma replied:
must have been a middle click by mistake… Shows how tests are important.
I'll probably add some loggers tests in the future in a separate PR.

borisdayma commented:
The last CI error seems unrelated to this PR.

jplu commented:
Ok it is fine for me for the TF part. I let @julien-c to review.

borisdayma commented:
Can I have any feedback on any possible changes still required?
This is a pretty large refactor of logging so it's hard to keep up to date with the repo.

borisdayma commented on src/transformers/trainer_utils.py:
```diff
@@ -1,6 +1,79 @@
+        WANDB_DISABLED:
+            (Optional): boolean - defaults to false, set to "true" to disable wandb entirely
+    """
+    if trainer.is_world_master():
```
Here we need to do one of the following:

* Option 1: add a `TFTrainer.is_world_master` method (I would need help for that)
* Option 2: replace with something like `if trainer.__class__.__name__ == "TFTrainer" or trainer.is_world_master()`

What do you think?

jplu replied:
No need of a `.is_world_master()` method in TF.

borisdayma replied:
Ok I moved it directly to `Trainer` so it does not apply to `TFTrainer`

borisdayma commented:
I merged master. Main change is logging for wandb & Tensorboard is applied only for world master now.
Need to figure out if the same should apply for `TFTrainer`. I left a comment on the code.

jplu commented:
Thanks! Looks ok, about the world master there is no need for the TF part.

jplu commented on src/transformers/trainer.py:
```diff
@@ -552,32 +514,24 @@ def train(self, model_path: Optional[str] = None):
+        for k, v in logs.items():
+            if isinstance(v, (int, float)):
+                self.tb_writer.add_scalar(k, v, self.global_step)
+            else:
```
In case the value is a string.

```suggestion
            elif isinstance(v, str):
                tf.summary.text(k, v, self.global_step)
            else:
```
Or even better you can cast the value to `str` to log it and log the error in case it cannot be cast.

jplu commented on src/transformers/trainer.py:
```diff
@@ -713,7 +667,8 @@ def evaluate(
         output = self._prediction_loop(eval_dataloader, description="Evaluation")
 
-        self._log(output.metrics)
+        if self.is_world_master:
```
To be removed.

jplu commented on src/transformers/trainer_utils.py:
```diff
@@ -1,6 +1,77 @@
+
+    wandb.init(project=os.getenv("WANDB_PROJECT", "huggingface"), config=vars(trainer.args))
+    # keep track of model topology and gradients (only with Pytorch)
+    if os.getenv("WANDB_WATCH") != "false" and trainer.__class__.__name__ == "Trainer":
```
Here instead of checking if the call comes from the PT or TF you should put this part into the PT trainer as it is not available in TF.

julien-c commented on src/transformers/trainer.py:
```diff
@@ -426,8 +387,8 @@ def train(self, model_path: Optional[str] = None):
-        self.global_step = 0
-        self.epoch = 0
+        self.global_step = 0  # used by loggers
+        self.epoch_logging = 0  # added with loggers to use as x-axis
```
nit: can you add comments on their own line (only if they're really needed)

jplu commented on src/transformers/trainer_utils.py:
```diff
@@ -1,6 +1,77 @@
+            trainer.model, log=os.getenv("WANDB_WATCH", "gradients"), log_freq=max(100, trainer.args.logging_steps)
+        )
+    # give access to wandb module
+    trainer._wandb = wandb
```
Better to return `wandb` than to set it.

julien-c commented on src/transformers/trainer_tf.py:
```diff
@@ -265,44 +285,40 @@ def train(self) -> None:
-            for training_loss in self._training_steps(train_ds, optimizer):
-                step = iterations.numpy()
+        self.global_step = 0  # used by loggers
+        self.epoch_logging = 0  # added with loggers to use as x-axis
```
same here

jplu commented on src/transformers/trainer_tf.py:
```diff
@@ -265,44 +285,40 @@ def train(self) -> None:
-        for epoch in range(start_epoch, int(epochs + 1)):
-            for training_loss in self._training_steps(train_ds, optimizer):
-                step = iterations.numpy()
+        self.global_step = 0  # used by loggers
```
The global init here can even be removed.

borisdayma replied:
It is used if you want to eval several pre-trained models without additional training.

I can also change the logging logic to check if this parameter is present at all.

jplu replied:
We already set it to 0 in the ``__init__`

julien-c commented on src/transformers/trainer_utils.py:
```diff
@@ -1,6 +1,77 @@
+    trainer._wandb = wandb
+
+
+def log_metrics(trainer, logs: Dict[str, float], iterator: Optional[tqdm] = None) -> None:
```
Not sure we want to share that accross both Trainer to be honest, this seems weird

borisdayma replied:
Are you talking about both the setup and logging part?

Initially I didn't notice tensorboard logging would be different between the 2 methods but now it may make sense to keep entire logging separate.

What about setting up wandb (I would just add a bool "watch" argument to make it clearer).

julien-c commented on src/transformers/trainer_utils.py:
```diff
@@ -1,6 +1,77 @@
+    if trainer.tb_writer:
+        trainer._log_tb(logs)
+    if is_wandb_available():
+        trainer._wandb.log(logs, step=trainer.global_step)
```
can't we just do `wandb.log`?

borisdayma commented:
Based on above comments, I can just put all the wandb logic separately in their respective Trainer's (and not use `trainer_utils` anymore).

Could you confirm this is the way to go @jplu @julien-c

borisdayma commented:
I made a new PR as I understood you prefer not to refactor logging into a single file like I did here.
Feel free to close this one if my understanding was correct.

borisdayma commented:
Obsolete PR
